File: tools/devonian/creatures/gemuendina/rework-v3/compare_interpolation_03.py
"""Bounded interpolation fixtures from actual exports plus edge-case glTF curves."""
import sys,json,struct,hashlib,collections
from pathlib import Path
import numpy as np
HERE=Path(__file__).resolve().parent;ROOT=HERE.parents[4];LOCAL=ROOT.parent/'devonian-authoring/gemuendina/rework-v3'
sys.dont_write_bytecode=True;sys.path.insert(0,str(HERE));from gltf_interpolation_03 import sample_channel
OUT=LOCAL/'interpolation-check-03';assert not OUT.exists();OUT.mkdir()
def sha(p):return hashlib.sha256(Path(p).read_bytes()).hexdigest()
cases=[];diagnostic=[]
for suffix in ('','.lod1'):
 p=LOCAL/'candidate-02'/('gemuendina'+suffix+'.glb');raw=p.read_bytes();n=struct.unpack_from('<I',raw,12)[0];g=json.loads(raw[20:20+n]);blob=raw[28+n:]
 def data(i):
  a=g['accessors'][i];v=g['bufferViews'][a['bufferView']];k={'SCALAR':1,'VEC3':3,'VEC4':4}[a['type']];assert a['componentType']==5126
  off=v.get('byteOffset',0)+a.get('byteOffset',0)
  return np.array([struct.unpack_from('<'+'f'*k,blob,off+j*v.get('byteStride',4*k))for j in range(a['count'])])
 counts=collections.Counter();step_constant=0
 for action in g['animations']:
  for ch in action['channels']:
   sp=action['samplers'][ch['sampler']];mode=sp.get('interpolation','LINEAR');times=data(sp['input']).ravel();values=data(sp['output']);counts[mode]+=1
   if mode=='STEP'and np.max(np.abs(values-values[0]))<1e-12:step_constant+=1
   cases.append({'origin':p.name+':'+action['name']+':'+g['nodes'][ch['target']['node']]['name'],
                 'mode':mode,'path':ch['target']['path'],'times':times.tolist(),'values':values.tolist()})
 diagnostic.append({'asset':str(p),'sha256':sha(p),'channel_modes':dict(counts),'constant_step_channels':step_constant})
# Nonconstant STEP, uneven key spacing, clamped endpoints, quaternion shortest
# path/near-coincident/half-turn, cubic dt scaling and normalized cubic rotation.
cases.extend([
 {'origin':'synthetic-changing-step','mode':'STEP','path':'translation','times':[.25,1.5,3.25],'values':[[0,1,2],[9,-3,4],[-2,8,1]]},
 {'origin':'synthetic-single-step','mode':'STEP','path':'translation','times':[.7],'values':[[1,2,3]]},
 {'origin':'synthetic-linear-vector','mode':'LINEAR','path':'translation','times':[.25,1.5,3.25],'values':[[0,1,2],[9,-3,4],[-2,8,1]]},
 {'origin':'synthetic-quaternion-sign','mode':'LINEAR','path':'rotation','times':[.25,1.5],'values':[[0,0,0,1],[0,0,0,-1]]},
 {'origin':'synthetic-quaternion-halfturn','mode':'LINEAR','path':'rotation','times':[.25,1.5],'values':[[0,0,0,1],[1,0,0,0]]},
 {'origin':'synthetic-quaternion-near','mode':'LINEAR','path':'rotation','times':[.25,1.5],'values':[[0,0,0,1],[.00001,0,0,.99999999995]]},
 {'origin':'synthetic-cubic-vector','mode':'CUBICSPLINE','path':'translation','times':[.25,2.75,4.0],'values':[[0,0,0],[1,2,3],[2,-3,1],[-1,2,3],[4,-2,1],[1,2,-1],[2,1,-1],[0,1,5],[0,0,0]]},
 {'origin':'synthetic-cubic-quaternion','mode':'CUBICSPLINE','path':'rotation','times':[.25,2.75],'values':[[0,0,0,0],[0,0,0,1],[.2,.1,0,0],[-.1,.05,0,.1],[.6,0,0,.8],[0,0,0,0]]}
])
for case in cases:
 case['times']=np.array(case['times'],dtype=np.float32).astype(float).tolist();case['values']=np.array(case['values'],dtype=np.float32).astype(float).tolist()
 ts=case['times'];queries=[ts[0]-.1,ts[-1]+.1]+ts
 for a,b in zip(ts,ts[1:]):queries.extend((a+(b-a)*.25,a+(b-a)*.5,a+(b-a)*.75,b-1e-7))
 case['queries']=sorted(set(queries));case['expected']=[sample_channel(ts,case['values'],t,case['mode'],case['path'])for t in case['queries']]
(OUT/'fixtures.json').write_text(json.dumps(cases,separators=(',',':')))
(OUT/'diagnostic.json').write_text(json.dumps({'helper_sha256':sha(HERE/'gltf_interpolation_03.py'),'assets':diagnostic,'cases':len(cases),'queries':sum(len(c['queries'])for c in cases)},indent=2)+'\n')
print('GEMUENDINA_INTERPOLATION_FIXTURES',len(cases),sum(len(c['queries'])for c in cases),OUT)

File: tools/devonian/creatures/gemuendina/rework-v3/compare_interpolation_03.mjs
/** Compare Python sampler results with the installed actual Three GLTFLoader. */
import fs from 'node:fs';import path from 'node:path';import {createHash}from'node:crypto';
import {GLTFLoader}from'three/examples/jsm/loaders/GLTFLoader.js';
const out=path.resolve('../devonian-authoring/gemuendina/rework-v3/interpolation-check-03');
const cases=JSON.parse(fs.readFileSync(path.join(out,'fixtures.json'),'utf8'));
if(fs.existsSync(path.join(out,'three-comparison.json')))throw Error('Preserve previous comparison');
const chunks=[],views=[],accessors=[],channels=[],samplers=[],nodes=[];let offset=0;
function accessor(rows,k){const flat=rows.flat(),buf=Buffer.alloc(flat.length*4);flat.forEach((v,i)=>buf.writeFloatLE(v,i*4));const index=accessors.length;views.push({buffer:0,byteOffset:offset,byteLength:buf.length});offset+=buf.length;chunks.push(buf);accessors.push({bufferView:views.length-1,componentType:5126,count:flat.length/k,type:k===1?'SCALAR':k===3?'VEC3':'VEC4',...(k===1?{min:[Math.min(...flat)],max:[Math.max(...flat)]}:{})});return index;}
for(const [i,c]of cases.entries()){nodes.push({name:'case_'+i});samplers.push({input:accessor(c.times,1),output:accessor(c.values,c.path==='rotation'?4:3),interpolation:c.mode});channels.push({sampler:i,target:{node:i,path:c.path}});}
const doc={asset:{version:'2.0'},scene:0,scenes:[{nodes:nodes.map((_,i)=>i)}],nodes,buffers:[{byteLength:offset}],bufferViews:views,accessors,animations:[{name:'fixture-curves',channels,samplers}]};
let json=Buffer.from(JSON.stringify(doc));json=Buffer.concat([json,Buffer.alloc((4-json.length%4)%4,32)]);const bin=Buffer.concat(chunks);const header=Buffer.alloc(20);header.writeUInt32LE(0x46546c67,0);header.writeUInt32LE(2,4);header.writeUInt32LE(28+json.length+bin.length,8);header.writeUInt32LE(json.length,12);header.writeUInt32LE(0x4e4f534a,16);const bh=Buffer.alloc(8);bh.writeUInt32LE(bin.length,0);bh.writeUInt32LE(0x004e4942,4);const glb=Buffer.concat([header,json,bh,bin]);
const loaded=await new GLTFLoader().parseAsync(glb.buffer.slice(glb.byteOffset,glb.byteOffset+glb.byteLength),'');
let maximum=0,queries=0;const failures=[];
for(const track of loaded.animations[0].tracks){const i=Number(track.name.split('.')[0].slice(5)),c=cases[i],interpolant=track.createInterpolant();
 for(const [j,t]of c.queries.entries()){const actual=Array.from(interpolant.evaluate(t)),expected=c.expected[j];let error=Math.max(...actual.map((v,k)=>Math.abs(v-expected[k])));if(c.path==='rotation')error=Math.min(error,Math.max(...actual.map((v,k)=>Math.abs(v+expected[k]))));maximum=Math.max(maximum,error);queries++;if(error>4e-6)failures.push({origin:c.origin,t,error,actual,expected});}}
const hash=p=>createHash('sha256').update(fs.readFileSync(p)).digest('hex');
const report={cases:cases.length,queries,maximum_component_error:maximum,tolerance:4e-6,failures,pass:failures.length===0,three_loader_sha256:hash('node_modules/three/examples/jsm/loaders/GLTFLoader.js'),fixture_sha256:hash(path.join(out,'fixtures.json')),comparison_source_sha256:hash(import.meta.filename)};
fs.writeFileSync(path.join(out,'three-comparison.json'),JSON.stringify(report,null,2)+'\n');if(failures.length)throw Error(JSON.stringify(failures.slice(0,3)));console.log('GEMUENDINA_INTERPOLATION_THREE_PASS',queries,maximum);
